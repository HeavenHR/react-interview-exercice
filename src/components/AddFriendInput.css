:local(.addFriendInput) {
  border-radius: 0;
  border-color: #abaaaa;
  border-left: 0;
  border-right: 0;
}

:local(.formContainer) {
  height: 40px;
}

:local(.genderDropdown) {
  height: 30px;
  width: 160px;
  margin: 5px 5px;
}

:local(.btnSubmit) {
  height: 30px;
  border-radius: 5px;
  margin-left: 15px;
}
