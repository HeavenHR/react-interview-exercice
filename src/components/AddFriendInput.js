import React, { Component } from "react";
import { PropTypes } from "prop-types";
import classnames from "classnames";
import styles from "./AddFriendInput.css";

class AddFriendInput extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      name: this.props.name || "",
      gender: ""
    };
  }

  handleChange(e) {
    this.setState({ name: e.target.value });
  }

  handleSelectChange(e) {
    this.setState({ gender: e.target.value });
  }

  handleSubmit(e) {
    e.preventDefault();
    const name = e.target[0].value.trim();
    const gender = e.target[1].value;
    this.props.addFriend(name, gender);
    this.setState({ name: "", gender: "" });
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit.bind(this)}>
        <input
          type="text"
          autoFocus="true"
          className={classnames("form-control", styles.addFriendInput)}
          placeholder="Type the name of a friend"
          value={this.state.name}
          onChange={this.handleChange.bind(this)}
        />
        <div className={styles.formContainer}>
          <select
            name="gender"
            value={this.state.gender}
            onChange={this.handleSelectChange.bind(this)}
            className={styles.genderDropdown}
          >
            <option value="">select gender</option>
            <option value="Male">Male</option>
            <option value="Female">Female</option>
          </select>
          <input
            type="submit"
            value="Add Friend"
            className={styles.btnSubmit}
          />
        </div>
      </form>
    );
  }
}

AddFriendInput.propTypes = {
  addFriend: PropTypes.func.isRequired
};

export default AddFriendInput;
