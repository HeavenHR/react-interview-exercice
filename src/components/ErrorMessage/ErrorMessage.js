import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

const ErrorMessage = ({ message }) =>
  <span className={classnames('help-block')}>{message}</span>;

ErrorMessage.propTypes = {
  message: PropTypes.string.isRequired
};

export default ErrorMessage;
