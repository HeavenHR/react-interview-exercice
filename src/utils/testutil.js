import { mount, shallow } from 'enzyme';
import { prettyPrint } from 'html';
import * as types from '../constants/ActionTypes';

export const mockFunction = jest.fn(value => value);

export const sampleFriend = {
  name: 'Sample friend',
  gender: 'male'
};

export const sampleId = 1;

export const createMountedComponent = ({ children, isShallow }) => {
  const mountedComponent = isShallow ? shallow(children) : mount(children);
  mountedComponent.htmlRender = () => prettyPrint(mountedComponent.html(), { indent_size: 2 });

  return mountedComponent;
};

export const inputProps = {
  value: 'sample name',
  onSubmit: mockFunction,
  onChange: mockFunction,
  isInvalid: false,
  errorMessage: 'Please fill friend name'
};

export const buttonGroupProps = {
  selected: '',
  options: [
    { value: 'male', showIcon: true },
    { value: 'female', showIcon: true }
  ],
  onClick: mockFunction,
  isInvalid: false,
  errorMessage: 'Please select gender'
};

export const changeEventObject = { target: { value: 'sample text' } };

export const samplePages = [1, 2, 3, 4, 5, 6, 7];

export const totalPages = samplePages.length;

export const addFriendAction = {
  type: types.ADD_FRIEND,
  friend: sampleFriend
};

export const expectedAddFriendState = {
  friendsById: [
    {
      name: 'Sample friend',
      gender: 'male'
    },
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const deleteFriendAction = {
  type: types.DELETE_FRIEND,
  id: 0
};

export const expectedDeleteFriendState = {
  friendsById: [
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const starFriendAction = {
  type: types.STAR_FRIEND,
  id: 1
};

export const expectedStarFriendState = {
  friendsById: [
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'male'
    },
    {
      name: 'Abraham Lincoln',
      starred: true,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export const updatePaginationAction = {
  type: types.UPDATE_PAGINATION,
  currentPage: 2
};

export const expectedUpdatedPaginationState = {
  pageSize: 2,
  currentPage: 2,
  pagingRange: 3
};
