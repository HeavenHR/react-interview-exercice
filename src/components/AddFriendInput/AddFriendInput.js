import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './AddFriendInput.css';
import ErrorMessage from '../ErrorMessage';

class AddFriendInput extends Component {
  static propTypes = {
    value: PropTypes.string,
    onSubmit: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    isInvalid: PropTypes.bool,
    errorMessage: PropTypes.string
  }

  render() {
    const { isInvalid, errorMessage } = this.props;
    return (
      <div className={classnames('form-group', { 'has-error': isInvalid })}>
        <input
          type="text"
          autoFocus="true"
          className={classnames('form-control', styles.addFriendInput)}
          placeholder="Type the name of a friend"
          value={this.state.value}
          onChange={this.handleChange.bind(this)}
          onKeyDown={this.handleSubmit.bind(this)}
        />
        {isInvalid && errorMessage && <ErrorMessage message={errorMessage} />}
      </div>
    );
  }

  constructor(props, context) {
    super(props, context);
    this.state = {
      value: this.props.value || ''
    };
  }

  componentWillReceiveProps(nextProps) {
    const { value } = nextProps;

    if (value !== this.props.value) {
      this.setState({ value });
    }
  }

  handleChange(e) {
    const value = e.target.value;
    const { onChange } = this.props;

    this.setState({ value });
    onChange && onChange(value);
  }

  handleSubmit(e) {
    const { onSubmit } = this.props;
    if (e.which === 13) {
      onSubmit && onSubmit();
    }
  }
}

export default AddFriendInput
