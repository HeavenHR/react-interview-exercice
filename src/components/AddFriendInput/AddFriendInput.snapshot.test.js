import React from 'react';

import AddFriendInput from './AddFriendInput';
import { createMountedComponent, inputProps, changeEventObject } from '../../utils/testutil';
import ErrorMessage from '../ErrorMessage';

const newText = 'new text';

describe('<AddFriendInput /> should', () => {
  let mountedAddFriendInput;
  let input;

  beforeEach(() => {
    mountedAddFriendInput = createMountedComponent({
      children: <AddFriendInput {...inputProps} />
    });

    input = mountedAddFriendInput.find('input');
  });

  it('render correctly according to props', () => {
    expect(mountedAddFriendInput.htmlRender()).toMatchSnapshot();
    expect(input.props().type).toEqual('text');
    expect(input.props().value).toEqual(inputProps.value);
    expect(input.props().className).toEqual('form-control');
  });

  it('render error message when has invalid prop', () => {
    const mountedAddFriendInput = createMountedComponent({
      children: <AddFriendInput {...inputProps} isInvalid={true} />
    });

    expect(mountedAddFriendInput.find('div').hasClass('has-error')).toBeTruthy();
    expect(mountedAddFriendInput.find(ErrorMessage)).toHaveLength(1);
  });

  it('updates state and call onChange function on input change', () => {
    input.simulate('change', changeEventObject);

    expect(mountedAddFriendInput.state().value).toEqual('sample text');
    expect(mountedAddFriendInput.props().onChange).toBeCalledWith('sample text');
  });

  it('call onSubmit function on press enter key', () => {
    input.simulate('keyDown', { which: 13 });

    expect(mountedAddFriendInput.props().onSubmit).toBeCalled();
  });
});
