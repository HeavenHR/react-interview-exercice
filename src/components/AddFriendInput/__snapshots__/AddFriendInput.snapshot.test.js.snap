exports[`<AddFriendInput /> should render correctly according to props 1`] = `
"<div class="form-group">
  <input type="text" class="form-control" placeholder="Type the name of a friend"
  value="sample name">
</div>"
`;
